packages/twenty-server/src/engine/api/graphql/graphql-query-runner/graphql-query-runner.service.ts

```diff
@@ -32,102 +36,240 @@ export class GraphqlQueryRunnerService {
       objectMetadataItem,
-      selectedFields,
+      graphqlFields(info),
+    );
+    const isForwardPagination = !isDefined(args.before);
+    const order = graphqlQueryParser.parseOrder(
+      args.orderBy ?? [],
+      isForwardPagination,
+    );
+    const where = graphqlQueryParser.parseFilter(
+      args.filter ?? ({} as Filter),
+      true,
     );
 
-    const order = args.orderBy
-      ? graphqlQueryParser.parseOrder(args.orderBy)
-      : undefined;
-
-    const where = args.filter
-      ? graphqlQueryParser.parseFilter(args.filter)
-      : {};
-
-    let cursor: Record<string, any> | undefined;
-
-    if (args.after) {
-      cursor = decodeCursor(args.after);
-    } else if (args.before) {
-      cursor = decodeCursor(args.before);
-    }
+    const cursor = this.getCursor(args);
+    const limit = this.getLimit(args);
 
-    if (args.first && args.last) {
-      throw new GraphqlQueryRunnerException(
-        'Cannot provide both first and last',
-        GraphqlQueryRunnerExceptionCode.ARGS_CONFLICT,
-      );
-    }
-
-    const take = args.first ?? args.last ?? QUERY_MAX_RECORDS;
+    this.addOrderByColumnsToSelect(order, select);
 
     const findOptions: FindManyOptions<ObjectLiteral> = {
       where,
       order,
       select,
       relations,
-      take,
+      take: limit + 1,
     };
-
-    const totalCount = await repository.count({
-      where,
-    });
+    const totalCount = await repository.count({ where });
 
     if (cursor) {
-      applyRangeFilter(where, order, cursor);
+      applyRangeFilter(where, cursor, isForwardPagination);
     }
 
     const objectRecords = await repository.find(findOptions);
+    const { hasNextPage, hasPreviousPage } = this.getPaginationInfo(
+      objectRecords,
+      limit,
+      isForwardPagination,
+    );
+
+    if (objectRecords.length > limit) {
+      objectRecords.pop();
+    }
 
     const typeORMObjectRecordsParser =
       new ObjectRecordsToGraphqlConnectionMapper(objectMetadataMap);
 
     return typeORMObjectRecordsParser.createConnection(
-      (objectRecords as ObjectRecord[]) ?? [],
-      take,
+      objectRecords as ObjectRecord[],
+      objectMetadataItem.nameSingular,
+      limit,
       totalCount,
       order,
-      objectMetadataItem.nameSingular,
+      hasNextPage,
+      hasPreviousPage,
     );
   }
+
+  private async getRepository(workspaceId: string, objectName: string) {
+    return this.twentyORMGlobalManager.getRepositoryForWorkspace(
+      workspaceId,
+      objectName,
+    );
+  }
+
+  private getObjectMetadata(
+    objectMetadataMap: Record<string, any>,
+    objectName: string,
+  ) {
+    const objectMetadata = objectMetadataMap[objectName];
+
+    if (!objectMetadata) {
+      throw new GraphqlQueryRunnerException(
+        `Object metadata not found for ${objectName}`,
+        GraphqlQueryRunnerExceptionCode.OBJECT_METADATA_NOT_FOUND,
+      );
+    }
+
+    return objectMetadata;
+  }
+
+  private getCursor(
```

Comment:
maybe move this to a pagination service to reduce the scope of graphql-query-runner.service